File: app/src/main/java/com/example/proyeksp/ui/RekeningViewModel.java
package com.example.proyeksp.ui;

import android.app.Application;
import android.net.Uri;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.example.proyeksp.database.Rekening;
import com.example.proyeksp.repository.RekeningRepo;

import java.util.List;

public class RekeningViewModel extends AndroidViewModel {
    private final RekeningRepo mRepository;
    private LiveData<List<Rekening>> rekeningList;

    public RekeningViewModel(Application application) {
        super(application);
        mRepository = new RekeningRepo(application);
        rekeningList = mRepository.getRekeningList();
    }
    public LiveData<Boolean> getSuccess() {
        return mRepository.getSuccess();
    }

    LiveData<List<Rekening>> getAllRekening() { return rekeningList; }
    public void update(Rekening rekening) {mRepository.update(rekening);}
    public Rekening getRekeningByNoRek(String s) {return mRepository.findRekeningByNoRek(s);}
    public void exportToXls(Uri uri) { mRepository.exportToXls(uri); }
    public void importFromXlsx(Uri uri) {
        try {
            mRepository.importFromXlsx(uri);
        } catch (RuntimeException e) {
            throw new RuntimeException(e);
        }
    }

    public LiveData<Integer> getScanData() { return mRepository.getScanData(); }
    public LiveData<Long> getTotalSetoran() {return mRepository.getTotalSetoran();}
}

File: app/src/main/java/com/example/proyeksp/ui/RekeningViewModel.kt
package com.example.proyeksp.ui

import android.app.Application
import android.net.Uri
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.asFlow
import androidx.lifecycle.viewModelScope
import com.example.proyeksp.database.Rekening
import com.example.proyeksp.database.Transaksi
import com.example.proyeksp.repository.RekeningRepo
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class RekeningViewModel(application: Application) : AndroidViewModel(application) {
    private val mRepository = RekeningRepo(application)
    val foundRekening = MutableLiveData<Rekening>()
    val scanNum = MutableLiveData<Int>()
    private val _allSetoran = MutableLiveData<List<Transaksi>>()
    val allSetoran: LiveData<List<Transaksi>> = _allSetoran
    val rekeningWithTodaySetoran: StateFlow<List<Rekening>> = mRepository.rekeningWithTodaySetoran

    val success = MutableLiveData<Boolean?>()

    init {
        fetchTransaksi()
    }

    fun resetSuccessState() {
        // Resetting to null prevents duplicate/stale Toasts from appearing
        success.value = null
    }

//    fun update(rekening: Rekening) {
//        mRepository.update(rekening)
//    }

//    fun getRekeningByNoRek(s: String): Rekening? {
//        return mRepository.findRekeningByNoRek(s)
//    }

    fun exportToXls(uri: Uri) {
        viewModelScope.launch {
            mRepository.exportToXls(uri)
        }
    }

//    fun importFromXlsx(uri: Uri) {
//        try {
//            mRepository.importFromXlsx(uri)
//        } catch (e: RuntimeException) {
//            throw RuntimeException(e)
//        }
//    }

//    val daftarRekening: LiveData<List<Rekening>>?
//        get() = mRepository.daftarRekening
//    val scanData: LiveData<Int>?
//        get() = mRepository.scanData
//    val totalSetoran: LiveData<Long?>?
//        get() = mRepository.totalSetoran

    // -------------------------------------------------------------------------------------

//    fun fetchAllRekening() {
//        viewModelScope.launch {
//            _allRekening.value = mRepository.getAllRekening()
//            Log.d("RekeningViewModel", "Fetched ${_allRekening.value?.size} records")
//        }
//    }

    fun getRekeningFromNoRek(s: String) {
        viewModelScope.launch {
            foundRekening.value = mRepository.getRekeningFromNoRek(s)
            Log.d("RekeningViewModel", "Found rekening: ${foundRekening.value}")
        }
    }

    fun updateRekening(rekening: Rekening) {
        viewModelScope.launch {
            mRepository.updateRekening(rekening)
        }
    }

    fun addSetoran(transaksi: Transaksi) {
        viewModelScope.launch {
            mRepository.addSetoran(transaksi)
        }
    }

    fun getScanData() {
        viewModelScope.launch {
            scanNum.value = mRepository.getNumberOfScan()
        }
    }

    fun fetchTransaksi() {
        viewModelScope.launch {
            mRepository.getRekeningWithTodaySetoran()
        }
    }
}
